src/Server/RemoteFSHandler.cpp

```diff
@@ -0,0 +1,543 @@
+            throw Exception(ErrorCodes::UNKNOWN_PACKET_FROM_CLIENT, "Unknown packet {} from client", toString(packet_type));
+    }
+}
+
+void RemoteFSHandler::receivePath(std::string & path)
+{
+    readStringBinary(path, *in);
+    LOG_TRACE(log, "Received path {}", path);
+}
+
+void RemoteFSHandler::receiveUnexpectedHello()
+{
+    String skip_string;
+
+    readStringBinary(skip_string, *in);
+
+    throw NetException(ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT, "Unexpected packet Hello received from client");
+}
+
+void RemoteFSHandler::iterateDirectory()
+{
+    std::string path;
+    receivePath(path);
+    for (auto iter = disk->iterateDirectory(path); iter->isValid(); iter->next())
+    {
+        LOG_TRACE(log, "Writing dir entry {}", iter->path());
+        writeVarUInt(RemoteFSProtocol::DataPacket, *out);
+        writeStringBinary(iter->path(), *out);
+    }
+    writeVarUInt(RemoteFSProtocol::EndIterateDirectory, *out);
+    out->next();
+}
+
+void RemoteFSHandler::listFiles()
+{
+    std::string path;
+    receivePath(path);
+    std::vector<String> files;
+    disk->listFiles(path, files);
+    writeVarUInt(files.size(), *out);
+    for (auto & file : files)
+    {
+        LOG_TRACE(log, "Writing file name {}", file);
+        writeVarUInt(RemoteFSProtocol::DataPacket, *out);
+        writeStringBinary(file, *out);
+    }
+    out->next();
+}
+
+void RemoteFSHandler::readFile()
+{
+    std::string str_data;
+    size_t offset;
+    size_t size;
+    receivePath(str_data);
+    readVarUInt(offset, *in);
+    LOG_TRACE(log, "Received offset {}", offset);
+    readVarUInt(size, *in);
+    LOG_TRACE(log, "Received size {}", size);
+
+    auto read_buf = disk->readFile(str_data);
+    read_buf->seek(offset, SEEK_SET);
+    str_data.resize(size);
+    auto bytes_read = read_buf->read(str_data.data(), size);
+    str_data.resize(bytes_read);
+
+    writeVarUInt(RemoteFSProtocol::ReadFile, *out);
+    writeStringBinary(str_data, *out);
+    out->next();
+}
+
+void RemoteFSHandler::writeFile()
+{
+    std::string str_data;
+    size_t buf_size;
+    uint mode_raw;
+    receivePath(str_data);
+    readVarUInt(buf_size, *in);
+    LOG_TRACE(log, "Received buf_size {}", buf_size);
+    readVarUInt(mode_raw, *in);
+    WriteMode mode = WriteMode(mode_raw);
+    LOG_TRACE(log, "Received mode {}", mode);
+
+    auto write_buf = disk->writeFile(str_data, buf_size, mode);
+    writeVarUInt(RemoteFSProtocol::StartWriteFile, *out);
+    out->next();
+    LOG_TRACE(log, "Sent {}", RemoteFSProtocol::StartWriteFile);
+
+    UInt64 packet_type;
+    while (true)
+    {
+        readVarUInt(packet_type, *in);
+        switch (packet_type)
+        {
+            case RemoteFSProtocol::DataPacket:
+                readStringBinary(str_data, *in);
+                LOG_TRACE(log, "Received data of size {}", str_data.size());
+                writeString(str_data, *write_buf);
+                write_buf->next(); // TODO maybe remove this line
```

Comment:
Looks like here `std_data` is the data for `writeFile` method and it can be unlimitedly big, we need to use `copyData()` method here (see `IO/copyData.h`).